File: pset6/vigenere.py
##
# Implements the vigenere cipher
##

import cs50
import sys


def main():

    # check if called with exactly one argument
    if len(sys.argv) != 2:
        show_usage()
        exit(1)

    # check if k consist only of alphabeticals
    k = sys.argv[1]
    k_len = len(k)
    if not k.isalpha():
        show_usage()
        exit(1)

    # ask for plaintext
    print("plaintext:  ", end="")
    p = cs50.get_string()

    # calculate and print encrypted text
    print("ciphertext: ", end="")
    i = 0
    for c in p:
        if c.isalpha:
            print(enc(ord(c), k[i]), end="")
            i += 1
            if i == k_len:
                i = 0
        else:
            print(c, end="")

    # print newline for better formatting
    print()
    exit(0)


# this function encrypts single character c with a key k
def enc(c, k):
    shift = 65
    if c > 96:
        shift = 97
    m = ord(k.upper()) - 65
    return chr(((c - shift + m) % 26) + shift)


# prints usage
def show_usage():
    print("Usage: ./vigenere k")

if __name__ == "__main__":
    main()

File: pset2/vigenere/vigenere.c
#include <stdio.h>
#include <cs50.h>
#include <string.h>
#include <ctype.h>

char enc(char c, char key);
void show_usage();

int main(int argc, string argv[])
{
    // check if called with exactly one argument
    if (argc != 2)
    {
        show_usage();
        return(1);
    }
    
    // get the key from the first argument
    string k = argv[1];
    int k_len = strlen(k);
    
    // check if k consist only of alphabeticals
    for (int i = 0; i < k_len; i++)
    {
        if (! isalpha(k[i]) )
        {
            show_usage();
        return(1);
        }
    }
    
    // ask for plaintext
    printf("plaintext:  ");
    string p = get_string();

    // calculate and print encrypted text
    printf("ciphertext: ");
    
    // this variable counts through the key
    int c = 0;
    for (int i = 0, l = strlen(p); i < l; i++)
    {

        // encrypt and print alphabetic characters
        if (isalpha(p[i]))
        {
            printf( "%c", enc(p[i],k[c]) );
            
            // cycle through the key
            c++;
            if (c == k_len)
            {
                c = 0;
            }
        }
        
        // just print everything else
        else
        {
            printf( "%c", p[i]);
        }
    }

    // print newline for bettter formatting
    printf("\n");
}

// this function encrypts single character c with a key k
char enc(char c, char k)
{
    // shift sets the distance to zero for the upper and the lower case letters
    char shift = 65;
    if (islower(c))
    {
        shift = 97;
    }
    
    // convert 'a' and 'A' to 0, 'b' and 'B' to 2 and so on
    k = toupper(k) - 65;
    
    // calculate and return result
    return ( ((c - shift + k) % 26) + shift  );
}

// print usage
void show_usage()
{
    printf("Usage: ./vigenere k\n");
}
